network/dag/tree/tree.go

```diff
@@ -0,0 +1,313 @@
+	t.orphanedLeaves = nil
+}
+
+func (t *tree) Load(leaves map[uint32][]byte) error {
+	// initialize tree
+	split := uint32(math.MaxUint32)
+	for k := range leaves {
+		if k < split {
+			split = k
+		}
+	}
+	t.resetDefaults(2 * split)
+
+	// build tree
+	// note: current implementation requires a maximum of d*2^d calls to Data.Add(), where d = t.Depth. Building the tree bottom up would require a max of 2^(d-1) calls.
+	var err error
+	var leaf []byte
+	for split, leaf = range leaves {
+		data := t.prototype.New()
+		err = data.UnmarshalBinary(leaf)
+		if err != nil {
+			return err
+		}
+		err = t.applyToPath(split, func(n *node) error {
+			return n.Data.Add(data)
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	// TODO: should leaves be dirty or not?
+	t.ResetUpdate()
+
+	return nil
+}
+
+func (t *tree) Insert(ref hash.SHA256Hash, clock uint32) error {
+	return t.applyToPath(clock, func(n *node) error {
+		return n.Data.Insert(ref)
+	})
+}
+
+// applyToPath calls fn on all nodes on the path from the root to leaf containing the clock value.
+// If the path/leaf does not exist it will be created.
+func (t *tree) applyToPath(clock uint32, fn func(n *node) error) error {
+	// grow tree if needed
+	for clock >= t.MaxSize {
+		t.reRoot()
+	}
+
+	// apply fn to all nodes on path
+	var current *node
+	next := t.Root
+	for next != nil {
+		current = next
+		err := fn(current)
+		if err != nil {
+			return fmt.Errorf("failed for node with splitLC %d: %w", next.SplitLC, err)
+		}
+		next = t.getNextNode(current, clock)
+	}
+	t.dirtyLeaves[current.SplitLC] = current
+
+	return nil
+}
+
+func (t *tree) newBranch(start, stop uint32) *node {
+	split := (stop + start) / 2
+	n := newNode(split, stop, t.prototype.New())
+	if stop-start > t.LeafSize {
+		n.Left = t.newBranch(start, split)
+	} else {
+		t.dirtyLeaves[n.SplitLC] = n
+	}
+	return n
+}
+
+// reRoot creates a new Root with Data from the current Root and adds current Root as its Left branch.
+func (t *tree) reRoot() {
+	newRoot := newNode(t.MaxSize, 2*t.MaxSize, t.Root.Data.Clone())
+	newRoot.Left = t.Root
+	t.Root = newRoot
+	t.MaxSize *= 2
+	t.Depth++
+}
+
+// getNextNode retrieves the next node based on the clock value. If the node does not exist it is created.
+func (t *tree) getNextNode(n *node, clock uint32) *node {
+	// return nil if n is a leaf
+	if n.isLeaf() {
+		return nil
+	}
+
+	if clock < n.SplitLC {
+		return n.Left
+	} else {
+		if n.Right == nil {
+			n.Right = t.newBranch(n.SplitLC, n.LimitLC)
+		}
+		return n.Right
+	}
+}
+
+func (t *tree) GetRoot() Data {
+	return t.Root.Data.Clone()
+}
+
+func (t *tree) GetZeroTo(clock uint32) (Data, uint32) {
+	data := t.Root.Data.Clone()
+	next := t.Root
+	for {
+		current := next
+		if clock < current.SplitLC {
+			if current.Right != nil {
+				// Only fails when Data structures do not match, which should not happen for Data managed by the tree.
+				_ = data.Subtract(current.Right.Data)
```

Comment:
The leaves contain `Data` of a single `page`, the parents contain the sum of their children, and the root contains sum of all leaves. This method returns the sum of all leaves prior to and including the leaf containing the requested `clock` value, so the sum of pages ZeroTo(page containing clock). This can be done by subtracting all pages to right of the requested range from the value in the tree root, which is what this method is doing.

For Xor subtraction is simply xor-ing again, but for iblts this a bit more complex.